### PR TITLE
add support for partial indexes

### PR DESCRIPTION
**Pull request by timgraham**
refs #104

### CODE REVIEW

File: django_mongodb/indexes.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from django.db.models.sql.query import Query
+
+
+def _get_condition_mql(self, model, schema_editor):
+    """Analogous to Index._get_condition_sql()."""
+    query = Query(model=model, alias_cols=False)
+    where = query.build_where(self.condition)
+    compiler = query.get_compiler(connection=schema_editor.connection)
+    mql_ = where.as_mql(compiler, schema_editor.connection)
+    # Transform aggregate() query syntax into find() syntax.
```
Comment (timgraham) [inline django_mongodb/indexes.py:11]: This stinks and doesn't work for multiple conditions / null conditions (but maybe could be done). I wonder if it would be better to have a separate path from `as_mql()` to generate `find()` compatible MQL rather than trying to convert from the aggregate syntax.

Comment (Jibola) [reply django_mongodb/indexes.py:11]: As I understand it, we're pulling the index conditions from a model's field definitions using the compiler built by the schema editor? Is this schema_editor compiler different from our Mongo one?

can self.condition ever be a list of items?


Speaking more broadly, I'd say, yes, let's include the `find` compatible conversions, but how many nodes/expressions would we need that for? Also, since `find` is much more limited there would still be gaps.

**Review by Jibola — COMMENTED** [review]
Added some comments! Will circle back and ask some more tomorrow :)

File: django_mongodb/schema.py [hunk 2]
```diff
@@ -122,11 +148,43 @@ def alter_index_together(self, model, old_index_together, new_index_together):
...
         if index.contains_expressions:
             return
+        kwargs = {}
+        filter_expression = defaultdict(dict)
+        if index.condition:
+            filter_expression.update(index._get_condition_mql(model, self))
+        if unique:
+            kwargs["unique"] = True
+            if field:
+                filter_expression[field.column].update({"$type": field.db_type(self.connection)})
```
[the remark this answers is not in the document]
Comment (timgraham) [reply django_mongodb/schema.py:177]: The comment is below: "Use partialFilterExpression to allow multiple null values for a unique constraint." It's not really in the best spot and should be amended to reference `$type` since partialFilterExpression is also used for `condition`.

File: django_mongodb/indexes.py (hunk 1 above)
Comment (timgraham) [reply django_mongodb/indexes.py:11]: `query.get_compiler()` is the same compiler we use for other queries. 

`condition` [is a `Q` object](https://docs.djangoproject.com/en/dev/ref/models/indexes/#condition): "For example, `condition=Q(pages__gt=400)` indexes records with more than 400 pages."

What doesn't currently work is multiple conditions like:
```python
Q(
    pub_date__gt=datetime.datetime(
        year=2015,
        month=1,
        day=1,
        tzinfo=timezone.get_current_timezone(),
    )
)
& Q(headline__contains="China")
```
Maybe that can be translated without too much difficulty. I didn't think about it too much.

Obviously, we can only support conditions that MongoDB supports.

**Comment by timgraham** [comment]
After e9711203ac720f34ce14ecb53489828887c4ef64, simple conditions no longer work, but we may need a separate path for generating this MQL anyway.
```
======================================================================
FAIL: test_boolean_restriction_partial (indexes.tests.PartialIndexTests.test_boolean_restriction_partial)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/tests/indexes/tests.py", line 453, in test_boolean_restriction_partial
    self.assertEqual(
AssertionError: "{'published': {'$eq': True}}" != "{'parent__field__0': {'$eq': True}}"
- {'published': {'$eq': True}}
+ {'parent__field__0': {'$eq': True}}
```

**Comment by WaVEV** [comment]
Sorry, misclicked. I will revert it.

**Comment by WaVEV** [comment]
> After [e971120](https://github.com/mongodb-labs/django-mongodb/commit/e9711203ac720f34ce14ecb53489828887c4ef64), simple conditions no longer work, but we may need a separate path for generating this MQL anyway.
> 
> ```
> ======================================================================
> FAIL: test_boolean_restriction_partial (indexes.tests.PartialIndexTests.test_boolean_restriction_partial)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/tim/code/django/tests/indexes/tests.py", line 453, in test_boolean_restriction_partial
>     self.assertEqual(
> AssertionError: "{'published': {'$eq': True}}" != "{'parent__field__0': {'$eq': True}}"
> - {'published': {'$eq': True}}
> + {'parent__field__0': {'$eq': True}}
> ```

Fixed, the columns are built without "alias" so the Col.as_mql fails.

File: django_mongodb/indexes.py (hunk 1 above)
Comment (WaVEV) [reply django_mongodb/indexes.py:11]: 🤔  we failed in the early stages when we used find instead of expr. 🤔 but maybe the indexes are different. It's not trivial to decide.

Comment (timgraham) [reply django_mongodb/indexes.py:11]: See https://www.mongodb.com/docs/manual/core/index-partial/#create-a-partial-index. It looks like it could be a subset of $find, so maybe implementing a separate code path is feasible and best.

Comment (WaVEV) [reply django_mongodb/indexes.py:11]: 🤔  I think it might be enough to re-implement the built-in lookup to use `find` instead of `expr`. This could be done with another resolver or a flag parameter. Lookups sometimes require extra parameters (not the built-in ones, but they exist); the downside is that those flags must be passed through the `where` node or injected when an expression is created. 

The `whereNode` should be adapted or a new one created, as `not`, `or`, and `and` are handled very differently.

Comment (timgraham) [reply django_mongodb/indexes.py:11]: That's what I was thinking. Something like how we had `as_mql` and `as_mql_agg`. I leave it to you.

**Comment by timgraham** [comment]
Can the following be supported? If not, we can edit the test with a condition that is supported.
```
======================================================================
ERROR: test_multiple_conditions (indexes.tests.PartialIndexTests.test_multiple_conditions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/test/testcases.py", line 1457, in skip_wrapper
    return test_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/tests/indexes/tests.py", line 486, in test_multiple_conditions
    sql = str(index._get_condition_mql(Article, schema_editor=editor))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/indexes.py", line 10, in _get_condition_mql
    return where.as_mql_idx(compiler, schema_editor.connection)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/query.py", line 316, in where_node_idx
    mql = child.as_mql_idx(compiler, connection)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/lookups.py", line 26, in builtin_lookup_idx
    return connection.mongo_operators_idx[self.lookup_name](lhs_mql, value)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
KeyError: 'contains'
```
Also, please add `tests/indexes_` in this repo so we can make sure we have complete test coverage for the new index MQL generation.

**Comment by timgraham** [comment]
Maybe it makes sense to put the `as_mql_idx` methods in `indexes.py`?

File: tests/indexes_/test_mql.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+from .models import Article
+
+
+class PartialIndexTests(TestCase):
+    def test_not_supported(self):
+        with connection.schema_editor() as editor:
+            index = Index(
+                name="not_supported",
+                # This is changed
```
Comment (timgraham) [inline tests/indexes_/test_mql.py:18]: You can remove this comment from these tests.

File: tests/indexes_/test_mql.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                    index.name,
+                    connection.introspection.get_constraints(
+                        cursor=cursor,
+                        table_name=Article._meta.db_table,
+                    ),
+                )
+            editor.remove_index(index=index, model=Article)
+
+    def test_raises_on_negated(self):
+        with self.assertRaises(NotSupportedError), connection.schema_editor() as editor:
```
Comment (timgraham) [inline tests/indexes_/test_mql.py:38]: Use `assertRaisesMessage` to verify the message as well.

**Review by timgraham — COMMENTED** [review]
Please make sure all operators in `mongo_operators_idx` are tested. Also, it looks like we may need quite a few tests to cover all the branches in `where_node_idx`.

File: django_mongodb/base.py [hunk 5]
```diff
@@ -119,6 +119,15 @@ def _isnull_operator(a, b):
         "iregex": lambda a, b: regex_match(a, b, insensitive=True),
     }
 
+    mongo_operators_idx = {
```
Comment (timgraham) [inline django_mongodb/base.py:122]: If this is all the operators we have, probably we don't need such a complex approach and can simply use a dict with pairs like` {"exact": "$eq"}` similar to `MONGO_OPERATORS` in `functions.py`.

File: tests/indexes_/models.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+class Article(models.Model):
+    headline = models.CharField(max_length=100)
+    pub_date = models.DateTimeField()
+    published = models.BooleanField(default=False)
+    number = models.IntegerField()
+    text = models.CharField(max_length=100)
+
+    class Meta:
+        indexes = [models.Index(fields=["headline", "pub_date"])]
```
Comment (timgraham) [inline tests/indexes_/models.py:12]: probably none of the tests require this index?

File: django_mongodb/features.py [hunk 7]
```diff
@@ -57,11 +57,6 @@ class DatabaseFeatures(BaseDatabaseFeatures):
         "db_functions.datetime.test_extract_trunc.DateFunctionWithTimeZoneTests.test_trunc_timezone_applied_before_truncation",
         # Length of null considered zero rather than null.
         "db_functions.text.test_length.LengthTests.test_basic",
-        # Partial indexes don't support multiple conditions. It requires this
-        # backend to convert $aggregate MQL syntax to $find (or else just
-        # generate $find syntax in the first place).
-        "indexes.tests.PartialIndexTests.test_is_null_condition",
```
Comment (timgraham) [inline django_mongodb/features.py]: I adjusted these tests on the Django fork.

**Review by timgraham — COMMENTED** [review]
This is coming along. I pushed some edits in lieu of leaving some comments.

File: tests/indexes_/models.py (hunk 6 above)
Comment (WaVEV) [reply tests/indexes_/models.py:12]: Yes, they aren't needed, forgot to remove.

File: django_mongodb/base.py (hunk 5 above)
Comment (WaVEV) [reply django_mongodb/base.py:122]: Yes, they are only a few much and fewer in mongo5, I will move this to indexes.py

File: django_mongodb/indexes.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    return where.as_mql_idx(compiler, schema_editor.connection)
+
+
+def builtin_lookup_idx(self, compiler, connection):
+    if not isinstance(self.lhs, Col):
+        raise NotSupportedError("Expressions as indexes are not supported.")
+    lhs_mql = self.lhs.target.column
+    value = process_rhs(self, compiler, connection)
+    try:
+        return mongo_operators_idx[self.lookup_name](lhs_mql, value)
```
Comment (WaVEV) [inline django_mongodb/indexes.py:36]: This implementation has a problem with tests constraints in Django, I think we have two options:

1. Modify the unit test  in order not to use `isnull` index. 
2. Change the isnull behaviour.

I think the option 1 is better.

Comment (timgraham) [reply django_mongodb/indexes.py:36]: I amended `test_is_null_condition` on the Django fork to reflect the fact that it raises `NotSupportedError`.

File: django_mongodb/indexes.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+from django.db.models.expressions import Col
+from django.db.models.fields.related_lookups import In
+from django.db.models.lookups import BuiltinLookup
+from django.db.models.sql.query import Query
+from django.db.models.sql.where import AND, XOR, WhereNode
+
+from .query_utils import process_rhs
+
+mongo_operators_idx = {
+    "exact": lambda a, b: {a: {"$eq": b}},
```
[the remark this answers is not in the document]
Comment (WaVEV) [reply django_mongodb/indexes.py:13]: Got it

**Comment by WaVEV** [comment]
Found a bug in where.as_mql (should rise EmptyResultSet), and also I have to research in which way the test from constraints will be adapted, it isn't straightforward

**Comment by timgraham** [comment]
Oh, I didn't notice the constraints problems you mentioned previously. I added a Django fork commit to place `isnulll` usage in `UniqueConstraint.condition` tests, but other problems like this remain which I think it what you're referring to:
```
======================================================================
ERROR: test_validate_expression_condition (constraints.tests.UniqueConstraintTests.test_validate_expression_condition)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tim/code/django/django/test/testcases.py", line 1457, in skip_wrapper
    return test_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/tests/constraints/tests.py", line 1014, in test_validate_expression_condition
    constraint.validate(UniqueConstraintProduct, non_unique_product)
  File "/home/tim/code/django/django/db/models/constraints.py", line 461, in validate
    if (self.condition & Exists(queryset.filter(self.condition))).check(
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/models/query_utils.py", line 140, in check
    return compiler.execute_sql(SINGLE) is not None
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 246, in execute_sql
    query = self.build_query(
            ^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 379, in build_query
    query.project_fields = self.get_project_fields(columns, ordering_fields)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 563, in get_project_fields
    fields.update(fields.pop(self.collection_name, {}))
                             ^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/utils/functional.py", line 47, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
                                         ^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 438, in collection_name
    base_table = next(
                 ^^^^^
StopIteration
```
The SQL:
```
SELECT 1 AS "_check"
WHERE COALESCE((EXISTS
                  (SELECT 1 AS "a"
                   FROM "constraints_uniqueconstraintconditionproduct" U0
                   WHERE (U0."name" = 'p1'
                          AND U0."color" IS NULL)
                   LIMIT 1)), 1);
```
I think it fails because in MongoDB, we have no notion of executing a query without a collection.

I wonder if we could make it work without the outer query? But it looks tricky because I don't see a place we could hook into to customize the query: https://github.com/django/django/blob/042b381e2e37c0c37b8a8f6cc9947f1a2ebfa0dd/django/db/models/query_utils.py#L117-L151

**Comment by WaVEV** [comment]
> Oh, I didn't notice the constraints problems you mentioned previously. I added a Django fork commit to place `isnulll` usage in `UniqueConstraint.condition` tests, but other problems like this remain which I think it what you're referring to:
> 
> ```
> ======================================================================
> ERROR: test_validate_expression_condition (constraints.tests.UniqueConstraintTests.test_validate_expression_condition)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/tim/code/django/django/test/testcases.py", line 1457, in skip_wrapper
>     return test_func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/code/django/tests/constraints/tests.py", line 1014, in test_validate_expression_condition
>     constraint.validate(UniqueConstraintProduct, non_unique_product)
>   File "/home/tim/code/django/django/db/models/constraints.py", line 461, in validate
>     if (self.condition & Exists(queryset.filter(self.condition))).check(
>        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/code/django/django/db/models/query_utils.py", line 140, in check
>     return compiler.execute_sql(SINGLE) is not None
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 246, in execute_sql
>     query = self.build_query(
>             ^^^^^^^^^^^^^^^^^
>   File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 379, in build_query
>     query.project_fields = self.get_project_fields(columns, ordering_fields)
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 563, in get_project_fields
>     fields.update(fields.pop(self.collection_name, {}))
>                              ^^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/code/django/django/utils/functional.py", line 47, in __get__
>     res = instance.__dict__[self.name] = self.func(instance)
>                                          ^^^^^^^^^^^^^^^^^^^
>   File "/home/tim/code/django-mongodb/django_mongodb/compiler.py", line 438, in collection_name
>     base_table = next(
>                  ^^^^^
> StopIteration
> ```
> 
> The SQL:
> 
> ```
> SELECT 1 AS "_check"
> WHERE COALESCE((EXISTS
>                   (SELECT 1 AS "a"
>                    FROM "constraints_uniqueconstraintconditionproduct" U0
>                    WHERE (U0."name" = 'p1'
>                           AND U0."color" IS NULL)
>                    LIMIT 1)), 1);
> ```
> 
> I think it fails because in MongoDB, we have no notion of executing a query without a collection.
> 
> I wonder if we could make it work without the outer query? But it looks tricky because I don't see a place we could hook into to customize the query: https://github.com/django/django/blob/042b381e2e37c0c37b8a8f6cc9947f1a2ebfa0dd/django/db/models/query_utils.py#L117-L151

I added like a dummy collection to handle this and created the pr fixing the remaining unit tests.
https://github.com/mongodb-forks/django/pull/11

File: django_mongodb/indexes.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    """Analogous to Index._get_condition_sql()."""
+    query = Query(model=model, alias_cols=False)
+    where = query.build_where(self.condition)
+    compiler = query.get_compiler(connection=schema_editor.connection)
+    return where.as_mql_idx(compiler, schema_editor.connection)
+
+
+def builtin_lookup_idx(self, compiler, connection):
+    if not isinstance(self.lhs, Col):
+        raise NotSupportedError("Expressions as indexes are not supported.")
```
Comment (timgraham) [inline django_mongodb/indexes.py:32]: Is this message tested? I don't see it in the new tests or in the Django test suite.

File: django_mongodb/indexes.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+            empty_needed -= 1
+        except FullResultSet:
+            full_needed -= 1
+        else:
+            if mql:
+                children_mql.append(mql)
+            else:
+                full_needed -= 1
+
+    if empty_needed == 0:
```
Comment (timgraham) [inline django_mongodb/indexes.py:81]: Could you carefully audit these branches? I added `assert False` to them and it doesn't appear test apps `constraints indexes indexes_` covers them. I'm not sure what it would mean to have an index where node that matches nothing or everything. Wouldn't that be kind of pointless? I guess as long as we match Django's behavior, it's okay.

File: django_mongodb/lookups.py [hunk 12]
```diff
@@ -48,6 +50,14 @@ def in_(self, compiler, connection):
 def is_null(self, compiler, connection):
     if not isinstance(self.rhs, bool):
         raise ValueError("The QuerySet value for an isnull lookup must be True or False.")
+    if isinstance(self.lhs, Value):
```
Comment (timgraham) [inline django_mongodb/lookups.py:53]: How did this come up in the context of this PR? What test does it fix?

File: django_mongodb/indexes.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+from django.db.models.sql.query import Query
+from django.db.models.sql.where import AND, XOR, WhereNode
+
+from .query_utils import process_rhs
+
+mongo_operators_idx = {
+    "exact": "$eq",
+    "gt": "$gt",
+    "gte": "$gte",
+    "lt": "$lt",
```
Comment (timgraham) [inline django_mongodb/indexes.py:16]: I think test coverage is missing for lt/lte.

File: django_mongodb/indexes.py (hunk 10 above)
Comment (WaVEV) [reply django_mongodb/indexes.py:32]: Will add.

File: django_mongodb/indexes.py (hunk 11 above)
Comment (WaVEV) [reply django_mongodb/indexes.py:81]: Yes, I will review carefully those edge cases and edit code or add test if it is needed.
I think there isn't a test in indexes that covers, I tried to imitate the other where.as_mql.

File: django_mongodb/lookups.py (hunk 12 above)
Comment (WaVEV) [reply django_mongodb/lookups.py:53]: Should be in a separate PR, I forgot to move the code. It fixed a test but in a side effect way.

File: django_mongodb/indexes.py (hunk 11 above)
Comment (WaVEV) [reply django_mongodb/indexes.py:81]: It isn't possible to create a case that covers that given the conditions of the indexes must be a Q expression, I did it in order to emulate the where.as_mql.  This part won't be used and the other try and catch expression also not needed.

File: django_mongodb/indexes.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        raise NotSupportedError(
+            f"MongoDB does not support the '{self.lookup_name}' lookup in indexes."
+        ) from None
+    return {lhs_mql: {operator: value}}
+
+
+def in_idx(self, compiler, connection):
+    if not connection.features.supports_in_index_operator:
+        raise NotSupportedError(
+            f"MongoDB < 6.0 does not support the '{self.lookup_name}' lookup in indexes."
```
Comment (Jibola) [inline django_mongodb/indexes.py:43]: Couldn't this just be:
```suggestion
            f"MongoDB < 6.0 does not support `in` lookup in indexes."
```

**Review by Jibola — APPROVED** [review]
Overall, LGTM. Would love clarity on the $facet question

File: django_mongodb/compiler.py [hunk 15]
```diff
@@ -435,11 +435,19 @@ def project_field(column):
...
+                for k, v in self.query.alias_map.items()
+                if isinstance(v, BaseTable) and self.query.alias_refcount[k]
+            )
+        except StopIteration:
+            # Use a dummy collection if the query doesn't specify a table
+            # (such as Constraint.validate() with a condition).
+            query = self.query_class(self)
+            query.aggregation_pipeline = [{"$facet": {"__null": []}}]
+            self.subqueries.insert(0, query)
+            return "__null"
```
[the remark this answers is not in the document]
Comment (WaVEV) [reply django_mongodb/compiler.py]: Well, the facet execution will have a very short pipeline, is just a way to insert a document so the subqueries won't get empty.
I mean
just this example:
```SQL
select 1
where exists(some subquery checking that an index exists)
```

This is translated as:
```JSON
[{"$facet": {"__null": []}},
 {"$lookup": {"the subquery"}}
]
```

If a query over an empty collection is executed, the result will be 0 rows. To avoid that, I put a single tiny document
